scheds/rust/scx_flash/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,1068 @@
+ */
+#define SLICE_MIN	(NSEC_PER_SEC / CONFIG_HZ)
+
+/*
+ * Task time slice range.
+ */
+const volatile u64 slice_max = 20ULL * NSEC_PER_MSEC;
+const volatile u64 slice_lag = 20ULL * NSEC_PER_MSEC;
+
+/*
+ * When enabled always dispatch all kthreads directly.
+ *
+ * This allows to prioritize critical kernel threads that may potentially slow
+ * down the entire system if they are blocked for too long, but it may also
+ * introduce interactivity issues or unfairness in scenarios with high kthread
+ * activity, such as heavy I/O or network traffic.
+ */
+const volatile bool local_kthreads;
+
+/*
+ * Scheduling statistics.
+ */
+volatile u64 nr_kthread_dispatches, nr_direct_dispatches, nr_shared_dispatches;
+
+/*
+ * Exit information.
+ */
+UEI_DEFINE(uei);
+
+/*
+ * CPUs in the system have SMT is enabled.
+ */
+const volatile bool smt_enabled = true;
+
+/*
+ * Current global vruntime.
+ */
+static u64 vtime_now;
+
+/*
+ * Per-CPU context.
+ */
+struct cpu_ctx {
+	struct bpf_cpumask __kptr *llc_mask;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__type(key, u32);
+	__type(value, struct cpu_ctx);
+	__uint(max_entries, 1);
+} cpu_ctx_stor SEC(".maps");
+
+/*
+ * Return a CPU context.
+ */
+struct cpu_ctx *try_lookup_cpu_ctx(s32 cpu)
+{
+	const u32 idx = 0;
+	return bpf_map_lookup_percpu_elem(&cpu_ctx_stor, &idx, cpu);
+}
+
+/*
+ * Per-task local storage.
+ *
+ * This contain all the per-task information used internally by the BPF code.
+ */
+struct task_ctx {
+	/*
+	 * Temporary cpumask for calculating scheduling domains.
+	 */
+	struct bpf_cpumask __kptr *llc_mask;
+
+	/*
+	 * Voluntary context switches metrics.
+	 */
+	u64 nvcsw;
+	u64 nvcsw_ts;
+	u64 avg_nvcsw;
+
+	/*
+	 * Task's average used time slice.
+	 */
+	u64 avg_runtime;
+	u64 sum_runtime;
+	u64 last_run_at;
+
+	/*
+	 * Task's deadline.
+	 */
+	u64 deadline;
+
+	/*
+	 * Task is holding a lock.
+	 */
+	bool lock_boost;
+};
+
+/* Map that contains task-local storage. */
+struct {
+	__uint(type, BPF_MAP_TYPE_TASK_STORAGE);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+	__type(key, int);
+	__type(value, struct task_ctx);
+} task_ctx_stor SEC(".maps");
+
+/*
+ * Return a local task context from a generic task.
+ */
+struct task_ctx *try_lookup_task_ctx(const struct task_struct *p)
+{
+	return bpf_task_storage_get(&task_ctx_stor,
+					(struct task_struct *)p, 0, 0);
+}
+
+/*
+ * User-space locking detection: re-using a logic similar to scx_lavd.
+ */
+struct futex_vector;
+struct hrtimer_sleeper;
+
+static bool is_task_locked(struct task_struct *p)
+{
+	struct task_ctx *tctx;
+
+	tctx = try_lookup_task_ctx(p);
+	return tctx ? tctx->lock_boost : false;
+}
+
+static void task_lock(struct task_struct *p)
+{
+	struct task_ctx *tctx;
+
+	tctx = try_lookup_task_ctx(p);
+	if (tctx)
+		tctx->lock_boost = true;
+}
+
+static void task_unlock(struct task_struct *p)
+{
+	struct task_ctx *tctx;
+
+	tctx = try_lookup_task_ctx(p);
+	if (tctx)
+		tctx->lock_boost = false;
```

Comment:
The current implementation does not seem to consider the case of holding multiple futexes.

Reply:
Overall very nice work!

Reply:
> The current implementation does not seem to consider the case of holding multiple futexes.

Correct, I'm just keeping things simple for now. And BTW, thanks for figuring out all these kprobes in lavd! :)